### PR TITLE
More optimization for cartesian subplots

### PR DESCRIPTION
**Pull request by etpinard**
to be more merged in https://github.com/plotly/plotly.js/pull/2474 

In progress for now, as I'll try to speed up a few more things in the coming days.

### CODE REVIEW

**Comment by etpinard** [comment]
I was hoping to improve a few more things in this PR, namely:

- optimize the number of [`positionLabels`](https://github.com/plotly/plotly.js/blob/f9a15be72e54562eb48a805e886f50947a55402a/src/plots/cartesian/axes.js#L1785-L1816) calls in `Axes.doTicks` which is called multiples time per axis
- speed up [`fixLabelOverlaps`](https://github.com/plotly/plotly.js/blob/f9a15be72e54562eb48a805e886f50947a55402a/src/plots/cartesian/axes.js#L1829-L1882)
- and maybe take a stab at https://github.com/plotly/plotly.js/issues/1988

but I'll have to leave that for later.

----

@alexcjohnson would mind taking a :eyes: at this?

File: src/plot_api/plot_api.js [hunk 1]
```diff
@@ -347,7 +347,7 @@ exports.plot = function(gd, data, layout, config) {
 
     // draw ticks, titles, and calculate axis scaling (._b, ._m)
     function drawAxes() {
-        return Axes.doTicks(gd, 'redraw');
+        return Axes.doTicks(gd, graphWasEmpty ? '' : 'redraw');
```
Comment (alexcjohnson) [inline src/plot_api/plot_api.js]: OK... this is the only caller using `doTicks(gd, '')` right? From the logic below it looks like there may have been one previously (or someone just calling `doTicks(gd)`) but I don't see it now. I'm a little reticent to add new branches inside `doTicks` when what we really need to do is unpack it into more digestible pieces - but that's probably a project for another time refactoring all of `axes.js`.

File: src/plot_api/subroutines.js [hunk 2]
```diff
@@ -124,31 +124,20 @@ exports.lsInner = function(gd) {
...
 
         if(overlappingDomain(xDomain, yDomain, lowerDomains)) {
-            plotgroupBgData = [0];
-        }
-        else {
+            var pgNode = plotgroup.node();
+            plotgroupBg = plotinfo.bg = Lib.ensureSingle(plotgroup, 'rect', 'bg');
+            pgNode.insertBefore(plotgroupBg.node(), pgNode.childNodes[0]);
+        } else {
+            plotgroupBg = plotgroup.select('rect,bg');
```
Comment (alexcjohnson) [inline src/plot_api/subroutines.js]: `rect.bg` perhaps?

File: src/plot_api/subroutines.js [hunk 3]
```diff
@@ -124,31 +124,20 @@ exports.lsInner = function(gd) {
...
+            if(plotgroupBg.size()) plotgroupBg.remove();
```
Comment (alexcjohnson) [inline src/plot_api/subroutines.js]: does this `if` make a difference? Does it take a non-negligible time to `remove` an empty selection?

File: src/plots/cartesian/axes.js [hunk 4]
```diff
@@ -2012,14 +2014,12 @@ axes.doTicks = function(gd, axid, skipTitle) {
...
         // configure their own titles.
-        var ax = axisIds.getFromId(gd, axid);
 
         // rangeslider takes over a bottom title so drop it here
         if(ax.rangeslider && ax.rangeslider.visible && ax._boundingBox && ax.side === 'bottom') return;
 
-        var avoidSelection = d3.select(gd).selectAll('g.' + axid + 'tick');
         var avoid = {
-            selection: avoidSelection,
+            selection: tickLabels,
```
Comment (alexcjohnson) [inline src/plots/cartesian/axes.js]: nice find!

File: src/plot_api/plot_api.js (hunk 1 above)
Comment (etpinard) [reply src/plot_api/plot_api.js]: > OK... this is the only caller using doTicks(gd, '') right? From the logic below it looks like there may have been one previously (or someone just calling doTicks(gd)) but I don't see it now.

That's correct.

File: src/plot_api/subroutines.js (hunk 2 above)
Comment (etpinard) [reply src/plot_api/subroutines.js]: Nice catch. I guess that means we have some untested code branches in there :rage:

Comment (alexcjohnson) [reply src/plot_api/subroutines.js]: > we have some untested code branches in there

maybe, maybe not. `rect,bg` would delete all `rect` elements (are there any others? if not, then what you have there will work, it just won't be as efficient) and all `bg` elements (which don't exist)

Comment (etpinard) [reply src/plot_api/subroutines.js]: fixed in 5307125ef9655bec0c2e24ba444b3cfc4c1c4f48

File: src/plot_api/subroutines.js (hunk 3 above)
Comment (etpinard) [reply src/plot_api/subroutines.js]: that `if` did not improve things -> :hocho: in 5307125ef9655bec0c2e24ba444b3cfc4c1c4f48

**Comment by alexcjohnson** [comment]
Fantastic- really whittling down the overhead! :dancer: :racehorse:

**Pull request merged**